File: iosApp/iosApp/Views/Consent/ConsentViewModel.swift
//
//  ConsentViewModel.swift
//  iosApp
//
//  Created by Jan Cortiel on 08.02.23.
//  Copyright © 2023 Ludwig Boltzmann Institute for
//  Digital Health and Prevention - A research institute
//  of the Ludwig Boltzmann Gesellschaft,
//  Oesterreichische Vereinigung zur Foerderung
//  der wissenschaftlichen Forschung
//  Licensed under the Apache 2.0 license with Commons Clause
//  (see https://www.apache.org/licenses/LICENSE-2.0 and
//  https://commonsclause.com/).
//

import AVFoundation
import Combine
import KMPNativeCoroutinesCombine
import UIKit
import shared

class ConsentViewModel: ObservableObject {
    private let coreModel: CoreConsentViewModel
    private let registration: RegistrationService
    var consentInfo: String?
    private let stringTable = "SettingsView"

    @Published var permissionModel: PermissionModel? = nil
    @Published var showErrorAlert: Bool = false
    @Published var requestedPermissions = false

    private var cancellables = Set<AnyCancellable>()

    lazy var permissionManager = PermissionManager()
    var permissionGranted = false

    init(registrationService: RegistrationService) {
        registration = registrationService
        coreModel = CoreConsentViewModel(registrationService: registrationService, studyConsentTitle: String(localized: "study_consent"))

        createPublisher(for: coreModel.permissions)
            .receive(on: DispatchQueue.main)
            .sink { _ in
            } receiveValue: { [weak self] model in
                self?.permissionModel = model
            }
            .store(in: &cancellables)
    }

    func onAppear() {
        permissionManager.observer = self
        coreModel.viewDidAppear()
    }

    func onDisappear() {
        permissionManager.observer = nil
        coreModel.viewDidDisappear()
    }

    func resetPermissionRequest() {
        requestedPermissions = false
        permissionManager.resetRequest()
    }

    func requestPermissions() {
        requestedPermissions = true
        permissionManager.requestPermission(permissionRequest: true)
    }

    private func acceptConsent() {
        if let uniqueId = UIDevice.current.identifierForVendor?.uuidString {
            registration.acceptConsent(uniqueDeviceId: uniqueId)
        }
    }

    func decline() {
        registration.declineConsent()
    }
}

extension ConsentViewModel: PermissionManagerObserver {
    func accepted() {
        Task { @MainActor in
            // Delegates to HealthConnectObservation's own collector-aware permission check (the
            // same logic already used at schedule-start time) instead of a hardcoded HealthKit
            // request here - it scopes to whichever subtypes the study actually needs, requests
            // only what's missing, and shows its own alert on decline.
            try? await AppDelegate.shared.observationFactory
                .observation(type: ObservationTypeEnum.healthConnect.value)?
                .updateObservationPermissions()
            if permissionManager.anyNeededPermissionDeclined() {
                AlertController.shared.openAlertDialog(
                    model:
                        AlertDialogModel.companion.fromStrings(
                            title: "Required Permissions Were Not Granted",
                            message: "This study requires one or more sensor permissions to function correctly. You may choose to decline these permissions; however, doing so may result in the application and study not functioning fully or as expected. Would you like to navigate to settings to allow the app access to these necessary permissions?",
                            confirmLabel: "Proceed to Settings",
                            cancelLabel: "Proceed Without Granting Permissions",
                            onConfirm: {
                                if let url = URL(string: UIApplication.openSettingsURLString), UIApplication.shared.canOpenURL(url) {
                                    UIApplication.shared.open(url, options: [:], completionHandler: nil)
                                }
                                self.resetPermissionRequest()
                            },
                            onDecline: {
                                self.acceptConsent()
                                self.requestedPermissions = false
                            }))
            } else {
                self.acceptConsent()
                self.requestedPermissions = false
            }
        }
    }
}

File: androidApp/src/main/java/io/redlink/umm/blendedcare/app/android/activities/consent/ConsentViewModel.kt
/*
 * Copyright LBI-DHP and/or licensed to LBI-DHP under one or more
 * contributor license agreements (LBI-DHP: Ludwig Boltzmann Institute
 * for Digital Health and Prevention -- A research institute of the
 * Ludwig Boltzmann Gesellschaft, Österreichische Vereinigung zur
 * Förderung der wissenschaftlichen Forschung).
 * Licensed under the Apache 2.0 license with Commons Clause
 * (see https://www.apache.org/licenses/LICENSE-2.0 and
 * https://commonsclause.com/).
 */
package io.redlink.umm.blendedcare.app.android.activities.consent

import android.content.Context
import androidx.lifecycle.ViewModel
import io.redlink.umm.blendedcare.app.android.BlendedCareApplication
import io.redlink.umm.blendedcare.app.android.R
import io.redlink.umm.blendedcare.app.android.extensions.getSecureID
import io.redlink.umm.blendedcare.app.android.extensions.stringResource
import io.redlink.umm.participant.AlertController
import io.redlink.umm.participant.models.AlertDialogModel
import io.redlink.umm.participant.registration.RegistrationService
import io.redlink.umm.participant.viewModels.permission.CoreConsentViewModel

class ConsentViewModel(
    val registrationService: RegistrationService
) : ViewModel() {
    val coreModel =
        CoreConsentViewModel(registrationService, stringResource(R.string.consent_information))

    fun acceptConsent(context: Context) {
        getSecureID(context)?.let { uniqueDeviceId ->
            registrationService.acceptConsent(uniqueDeviceId)
        }
    }

    fun openPermissionDeniedAlertDialog(context: Context) {
        AlertController.openAlertDialog(
            AlertDialogModel(
                title = stringResource(R.string.required_permissions_not_granted_title),
                message = stringResource(R.string.required_permission_not_granted_message),
                confirmLabel = stringResource(R.string.proceed_to_settings_button),
                cancelLabel = stringResource(R.string.proceed_without_granting_button),
                onConfirm = {
                    BlendedCareApplication.openSettings.value = true
                },
                onDecline = {
                    acceptConsent(context)
                }
            ))
    }

    fun openNotificationPermissionDeniedAlertDialog(context: Context) {
        AlertController.openAlertDialog(
            AlertDialogModel(
                title = stringResource(R.string.notification_permission_not_granted_title),
                message = stringResource(R.string.notification_permission_not_granted_message),
                confirmLabel = stringResource(R.string.proceed_to_settings_button),
                cancelLabel = stringResource(R.string.proceed_without_granting_button),
                onConfirm = {
                    BlendedCareApplication.openSettings.value = true
                },
                onDecline = {
                    acceptConsent(context)
                }
            ))
    }

    fun decline() {
        registrationService.declineConsent()
    }
}
